test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,505 @@
+    end)
+    luatest.assert(ok, string.format(
+        'error while demoting follower with elections on: %s', err))
+end
+
+-- Promoting current raft leader should succeed,
+-- even if he doesn't own synchro queue with elections enabled.
+g_nspaces.test_raft_leader_promote = function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+
+    -- Promote server, but get stuck before obtaining synchro queue
+    -- (write term bump to wal, get stuck on promote)
+    wal_delay_start(g.server_1, 2)
+    g.server_1:exec(function()
+        box.ctl.promote()
+    end)
+    g.server_1:wait_wal_delay()
+    g.server_1:wait_election_state('leader')
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+
+    luatest.assert(ok, string.format(
+        'error while promoting raft leader: %s', err))
+end
+
+g_nspaces.after_test('test_raft_leader_promote', function(g)
+    -- Finish promoting server_1
+    wal_delay_end(g.server_1)
+    g.server_1:wait_synchro_queue_owner_id()
+
+    g.server_1:box_config({election_mode = 'off'})
+    g.cluster:demote(g.server_1)
+    g.server_2:wait_synchro_queue_owner_id(0)
+end)
+
+-- Promoting and demoting should work when everything is ok.
+g_nspaces.test_ok = function(g)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    g.server_2:wait_synchro_queue_owner_id(g.server_1:instance_id())
+    luatest.assert(ok, string.format(
+        'error while promoting with elections off: %s', err))
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    g.server_2:wait_synchro_queue_owner_id(0)
+    luatest.assert(ok, string.format(
+        'error while demoting with elections off: %s', err))
+
+    g.server_1:box_config({election_mode = 'manual'})
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    g.server_2:wait_synchro_queue_owner_id(g.server_1:instance_id())
+    luatest.assert(ok, string.format(
+        'error while promoting with elections on: %s', err))
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    g.server_2:wait_lsn(g.server_1)
+    luatest.assert(ok, string.format(
+        'error while demoting with elections on: %s', err))
+end
+
+g_nspaces.after_test('test_ok', function(g)
+    -- Demoting server with election_mode ~= 'off' leaves it as limbo owner.
+    -- https://github.com/tarantool/tarantool/issues/6860
+    g.server_1:box_config({election_mode = 'off'})
+    g.cluster:promote(g.server_2)
+    g.cluster:demote(g.server_2)
+    g.server_1:wait_synchro_queue_owner_id(0)
+end)
+
+-- Simultaneous promoting/demoting should fail.
+g_nspaces.test_simultaneous = function(g)
+    wal_delay_start(g.server_1)
+
+    local wal_write_count = g.server_1:wal_write_count()
+    g.f = promote_start(g.server_1)
+    g.server_1:wait_wal_write_count(wal_write_count + 1)
```

Comment:
Are you waiting for a term bump here? Then I would suggest to wait for it explicitly. WAL write count is a fragile thing when it comes to procedures which might make multiple WAL writes.